webapp/components/utils/Notifications.js

```diff
@@ -40,4 +40,140 @@ export const notifications = [
+  let isComment = false
+  let isReport = false
+  let reasonTranslationExtention = ''
+  let iconName, iconTooltip, triggerer, title, author, linkName, linkParams, linkHashParam
+
+  // extract data out of the deep structure of db response
+
+  // leave undefined data as default, see above, so later by priority user, comment, post we get easely a clou what it is
+  switch (from.__typename) {
+    case 'Comment':
+      comment = from
+      post = comment.post
+      isComment = true
+      iconName = 'comment'
+      iconTooltip = 'notifications.comment'
+      triggerer = comment.author
+      break
+    case 'Post':
+      post = from
+      isPost = true
+      iconName = 'bookmark'
+      iconTooltip = 'notifications.post'
+      triggerer = post.author
+      break
+    case 'Report':
+      {
+        const filed = from.filed[0] // for readability
+        report = {
+          reasonCategory: filed.reasonCategory,
+          reasonDescription: filed.reasonDescription,
+        }
+        isReport = true
+        iconName = 'balance-scale'
+        iconTooltip = 'notifications.report.name'
+        triggerer = currentUser
+        switch (filed.reportedResource.__typename) {
+          case 'User':
+            user = filed.reportedResource
+            isUser = true
+            reasonTranslationExtention = '.user'
+            break
+          case 'Comment':
+            comment = filed.reportedResource
+            post = filed.reportedResource.post
+            isComment = true
+            reasonTranslationExtention = '.comment'
+            break
+          case 'Post':
+            post = filed.reportedResource
+            isPost = true
+            reasonTranslationExtention = '.post'
+            break
+        }
+      }
+      break
+  }
+
+  // extract the content and link data by priority: user, comment, post
+  if (user) {
+    // it's a user
+    title = user.name
+    author = user
+    linkName = 'profile-id-slug'
+    linkParams = { id: user.id, slug: user.slug }
+  } else {
+    // it's a comment or post
+    title = post.title
+    linkName = 'post-id-slug'
+    linkParams = { id: post.id, slug: post.slug }
+    if (comment) {
+      // it's a comment
+      author = comment.author
+      contentExcerpt = comment.contentExcerpt
+      linkHashParam = { hash: `#commentId-${comment.id}` }
+    } else {
+      // it's a post
+      author = post.author
+      contentExcerpt = post.contentExcerpt
+      linkHashParam = {}
+    }
+  }
+
+  const data = {
+    createdAt: notification.createdAt,
+    read: notification.read,
+    reason: notification.reason,
+    notificationSourceId: from.id,
+    isUser,
+    isPost,
+    isComment,
+    isReport,
+    iconName,
+    iconTooltip,
+    triggerer,
+    user,
+    comment,
+    post,
+    author,
+    title,
+    contentExcerpt,
+    report,
+    reasonTranslationExtention,
+    linkTo: { name: linkName, params: linkParams, ...linkHashParam },
+  }
+  return data
+}
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-01-23T19:36:46Z" title="Thursday, January 23rd 2020, 8:36:46 pm +01:00">Jan 23, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -40,4 +40,117 @@ export const notifications = [
       },
     },
   },
+  {
+    read: false,
+    reason: 'filed_report_on_resource',
+    from: {
+      __typename: 'Report',
+      id: 'reportOnUser',
+      filed: [
+        {
+          reasonCategory: 'discrimination_etc',
+          reasonDescription: 'This user is harassing me with bigoted remarks!',
+          reportedResource: {
+            __typename: 'User',
+            id: 'badWomen',
+            slug: 'mrs.-badwomen',
+            name: 'Mrs. Badwomen',
+          },
+        },
+      ],
+    },
+  },
 ]
+
+export const extractNotificationDataOfCurrentUser = (notification, currentUser) => {
+  const from = notification.from
+  let triggerer
+  const id = from.id
+  const createdAt = notification.createdAt
+  const read = notification.read
+  const reason = notification.reason
+  let title
+  let author
+  let user = null
+  let post = null
+  let comment = null
+  let contentExcerpt = null
+  let report = null
+  let reasonExtention = ''
+
+  if (from.__typename === 'Post') {
+    post = from
+    triggerer = post.author
+  } else if (from.__typename === 'Comment') {
+    comment = from
+    triggerer = comment.author
+    post = comment.post
+  } else if (from.__typename === 'Report') {
+    report = {
+      reasonCategory: from.filed[0].reasonCategory,
+      reasonDescription: from.filed[0].reasonDescription,
+    }
+    triggerer = currentUser
+    if (from.filed[0].reportedResource.__typename === 'User') {
+      user = from.filed[0].reportedResource
+      reasonExtention = '.user'
+    } else if (from.filed[0].reportedResource.__typename === 'Post') {
+      post = from.filed[0].reportedResource
+      reasonExtention = '.post'
+    } else if (from.filed[0].reportedResource.__typename === 'Comment') {
+      comment = from.filed[0].reportedResource
+      post = from.filed[0].reportedResource.post
+      reasonExtention = '.comment'
+    }
+  }
+
+  if (user) {
+    title = user.name
+    author = user
+  } else {
+    title = post.title
+    if (comment) {
+      author = comment.author
+      contentExcerpt = comment.contentExcerpt
+    } else {
+      author = post.author
+      contentExcerpt = post.contentExcerpt
+    }
+  }
+
+  const params = user
+    ? {
+        id: user.id,
+        slug: user.slug,
+      }
+    : post
+    ? {
+        id: post.id,
+        slug: post.slug,
+      }
+    : {}
+  const hashParam = comment ? { hash: `#commentId-${comment.id}` } : {}
+  const linkTo = {
+    name: user ? 'profile-id-slug' : 'post-id-slug',
+    params,
+    ...hashParam,
+  }
+
+  const data = {
+    triggerer,
+    id,
+    createdAt,
+    read,
+    reason,
+    title,
+    user,
+    post,
+    comment,
+    contentExcerpt,
+    author,
+    report,
+    reasonExtention,
+    linkTo,
+  }
+  return data
+}
```


wow! this is complicated! I gotta believe most of this is the backends job, oder?
I'd have to play around with it/wait for other people's opinions, but seems like the frontend is doing a lot of extra work here :worried:

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2020-01-24T09:13:15Z" title="Friday, January 24th 2020, 10:13:15 am +01:00">Jan 24, 2020</time>_

---


Hm is it so complicated?
A lot of this code was in the Vue components directly before. And because the hierarchy gets deeper with the `Report`notification …
… it extracts the data into a flat hierarchy.
This way the Vue components can be much less complex.

If you put it some way in the backend then it will look the same and you have kind of logic there that is only related to the way we display the data in the frontend.

I could write some comments to give a little hint of the ideas of this extraction? @mattwr18 

I’m curious to other ones opinions @Human-Connection/hc-dev-team ?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-02-01T14:50:50Z" title="Saturday, February 1st 2020, 3:50:50 pm +01:00">Feb 1, 2020</time>_

---


Yes! Too complicated. I didn't understand enough to tell if this is the job of the backend or the frontend though.

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2020-02-05T11:45:58Z" title="Wednesday, February 5th 2020, 12:45:58 pm +01:00">Feb 5, 2020</time>_

---


I would like to discuss this personal with you. @mattwr18